src/hpm.c

```diff
@@ -1,350 +1,345 @@
+        val &= ~bitmask;                                                       \
+        val |= bitmask;                                                        \
+        __asm__ __volatile__("csrw mhpmevent" #x ", %0" : : "r"(val));         \
+        break;
+
+/* Macro to set values into event selector register */
+#define METAL_HPM_CLR_EVENT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        val &= ~bitmask;                                                       \
+        __asm__ __volatile__("csrw mhpmevent" #x ", %0" : : "r"(val));         \
+        break;
+
+/* Macro to get values from event selector register */
+#define METAL_HPM_GET_EVENT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        break;
+
+/* Macro to read HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HPM_GET_COUNT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        do {                                                                   \
+            __asm__ __volatile__("csrr %0, mhpmcounter" #x "h" : "=r"(vh));    \
+            __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));        \
+            __asm__ __volatile__("csrr %0, mhpmcounter" #x "h" : "=r"(vh1));   \
+        } while (vh != vh1);                                                   \
+        break;
+#else
+#define METAL_HPM_GET_COUNT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \
+        break;
+#endif
+
+/* Macro to clear HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HPM_CLR_COUNT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrw mhpmcounter" #x "h, zero");                 \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        __asm__ __volatile__("csrw mhpmcounter" #x "h, zero");                 \
+        break;
+#else
+#define METAL_HPM_CLR_COUNT_REG(x)                                             \
+    case METAL_HPM_COUNTER_##x:                                                \
+        __asm__ __volatile__("csrw mhpmcounter" #x ", zero");                  \
+        break;
+#endif
+
+/* Max available counters */
+#define METAL_HPM_COUNT_MAX 32
+
+/* Macro to check for instruction trap */
+#define MCAUSE_ILLEGAL_INST 0x02
```

Comment:
Should use METAL_II_EXCEPTION_CODE instead. But this can be done in following commit